.github/workflows/release.yml

```diff
@@ -4,21 +4,24 @@ on:
     tags:
       - "*"
 
-permissions:
-  contents: read
+permissions: {}
 
 jobs:
   build:
     name: Build artifacts
+    permissions:
+      contents: read
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v6
+      - uses: actions/checkout@8e8c483db84b4bee98b60c0593521ed34d9990e8  # v6.0.1
         with:
-          fetch-depth: 0
-      - uses: denoland/setup-deno@v2
+          fetch-depth: 0  # Needed for hatch-vcs versioning
+          persist-credentials: false
+      - uses: denoland/setup-deno@e95548e56dfa95d4e1a28d6f422fafe75c4c26fb  # v2.0.3
         with:
           deno-version: v2.x
-      - uses: actions/setup-python@v6
+          cache: false
+      - uses: actions/setup-python@83679a892e2d95755f2dac6acb0bfd1e9ac5d548  # v6.1.0
```

Comment:
We will have to think about this, see <https://api.github.com/repos/actions/python-versions/releases/tags/3.15.0-alpha.2-19505847594> the release is not immutable.